File: scripts/gguf.py
"""Minimal GGUF reader: metadata, tensor directory, and dequantisation of the
types small Llama checkpoints use (F32, F16, Q8_0, Q4_0, Q4_K, Q6_K).

It exists so the reference model and the tests depend on nothing but numpy
and the file. Layout follows ggml's `gguf.c` and `ggml-quants.c`.
"""

import mmap
import pathlib
import struct
import json

import numpy as np

GGML_TYPES = {
    0: ("F32", 1, 4),
    1: ("F16", 1, 2),
    2: ("Q4_0", 32, 18),
    8: ("Q8_0", 32, 34),
    12: ("Q4_K", 256, 144),
    14: ("Q6_K", 256, 210),
    30: ("BF16", 1, 2),
}

_SCALAR = {0: "<B", 1: "<b", 2: "<H", 3: "<h", 4: "<I", 5: "<i", 6: "<f", 7: "<?", 10: "<Q", 11: "<q", 12: "<d"}


class GGUF:
    def __init__(self, path):
        self.path = pathlib.Path(path)
        f = open(self.path, "rb")
        self.buf = mmap.mmap(f.fileno(), 0, access=mmap.ACCESS_READ)
        self.pos = 0
        assert self._read("4s") == b"GGUF", "not a GGUF file"
        self.version = self._read("<I")
        n_tensors, n_kv = self._read("<Q"), self._read("<Q")
        self.meta = {}
        for _ in range(n_kv):
            k = self._str()
            self.meta[k] = self._value(self._read("<I"))
        self.tensors = {}
        for _ in range(n_tensors):
            name = self._str()
            nd = self._read("<I")
            dims = [self._read("<Q") for _ in range(nd)]
            ty = self._read("<I")
            off = self._read("<Q")
            self.tensors[name] = (dims, ty, off)
        align = self.meta.get("general.alignment", 32)
        self.data = (self.pos + align - 1) // align * align

    def _read(self, fmt):
        (v,) = struct.unpack_from(fmt, self.buf, self.pos)
        self.pos += struct.calcsize(fmt)
        return v

    def _str(self):
        n = self._read("<Q")
        s = bytes(self.buf[self.pos : self.pos + n]).decode("utf-8", errors="replace")
        self.pos += n
        return s

    def _value(self, t):
        if t in _SCALAR:
            return self._read(_SCALAR[t])
        if t == 8:
            return self._str()
        if t == 9:
            et, n = self._read("<I"), self._read("<Q")
            return [self._value(et) for _ in range(n)]
        raise ValueError(f"unknown gguf value type {t}")

    def type_name(self, name):
        return GGML_TYPES.get(self.tensors[name][1], (f"type{self.tensors[name][1]}",))[0]

    def tensor(self, name):
        """Dequantised to float32, shaped [rows, cols] (ggml's ne reversed)."""
        dims, ty, off = self.tensors[name]
        tname, blk, bsz = GGML_TYPES[ty]
        n = int(np.prod(dims))
        nbytes = n // blk * bsz
        raw = np.frombuffer(self.buf, np.uint8, nbytes, self.data + off)
        x = DEQUANT[tname](raw, n)
        return x.reshape(list(reversed(dims)))


def _f32(raw, n):
    return raw.view("<f4").astype(np.float32)


def _f16(raw, n):
    return raw.view("<f2").astype(np.float32)


def _bf16(raw, n):
    return (raw.view("<u2").astype(np.uint32) << 16).view(np.float32)


def _q8_0(raw, n):
    b = raw.reshape(-1, 34)
    d = b[:, :2].copy().view("<f2").astype(np.float32)
    q = b[:, 2:].view(np.int8).astype(np.float32)
    return (d * q).reshape(-1)


def _q4_0(raw, n):
    b = raw.reshape(-1, 18)
    d = b[:, :2].copy().view("<f2").astype(np.float32)
    qs = b[:, 2:]
    lo = (qs & 0x0F).astype(np.int8) - 8
    hi = (qs >> 4).astype(np.int8) - 8
    return (d * np.concatenate([lo, hi], axis=1).astype(np.float32)).reshape(-1)


def _k_scale_min(scales):
    """Unpack Q4_K's 12 bytes of 6-bit scales and mins (8 each)."""
    s = scales.astype(np.uint8)
    sc = np.empty((s.shape[0], 8), np.uint8)
    mn = np.empty((s.shape[0], 8), np.uint8)
    sc[:, :4] = s[:, 0:4] & 63
    mn[:, :4] = s[:, 4:8] & 63
    sc[:, 4:] = (s[:, 8:12] & 0x0F) | ((s[:, 0:4] >> 6) << 4)
    mn[:, 4:] = (s[:, 8:12] >> 4) | ((s[:, 4:8] >> 6) << 4)
    return sc.astype(np.float32), mn.astype(np.float32)


def _q4_k(raw, n):
    b = raw.reshape(-1, 144)
    d = b[:, 0:2].copy().view("<f2").astype(np.float32)
    dmin = b[:, 2:4].copy().view("<f2").astype(np.float32)
    sc, mn = _k_scale_min(b[:, 4:16])
    qs = b[:, 16:144].reshape(-1, 4, 32)  # 4 chunks of 64 values
    out = np.empty((b.shape[0], 8, 32), np.float32)
    out[:, 0::2] = qs & 0x0F
    out[:, 1::2] = qs >> 4
    out = d[:, :, None] * sc[:, :, None] * out - dmin[:, :, None] * mn[:, :, None]
    return out.reshape(-1)


def _q6_k(raw, n):
    b = raw.reshape(-1, 210)
    ql = b[:, 0:128].reshape(-1, 2, 64)
    qh = b[:, 128:192].reshape(-1, 2, 32)
    sc = b[:, 192:208].view(np.int8).astype(np.float32).reshape(-1, 2, 8)
    d = b[:, 208:210].copy().view("<f2").astype(np.float32)
    out = np.empty((b.shape[0], 2, 128), np.float32)
    for half in range(2):
        l, h = ql[:, half].astype(np.int32), qh[:, half].astype(np.int32)
        q1 = (l[:, 0:32] & 0xF) | (((h >> 0) & 3) << 4)
        q2 = (l[:, 32:64] & 0xF) | (((h >> 2) & 3) << 4)
        q3 = (l[:, 0:32] >> 4) | (((h >> 4) & 3) << 4)
        q4 = (l[:, 32:64] >> 4) | (((h >> 6) & 3) << 4)
        q = np.concatenate([q1, q2, q3, q4], axis=1) - 32  # 128 values
        s = np.repeat(sc[:, half], 16, axis=1)  # one scale per 16
        out[:, half] = d * s * q
    return out.reshape(-1)


DEQUANT = {"F32": _f32, "F16": _f16, "BF16": _bf16, "Q8_0": _q8_0, "Q4_0": _q4_0, "Q4_K": _q4_k, "Q6_K": _q6_k}


def ollama_model(name, root=pathlib.Path.home() / ".ollama/models"):
    """Path of the GGUF blob behind an Ollama model tag, e.g. 'llama3.2:1b'."""
    repo, tag = (name.split(":") + ["latest"])[:2]
    m = json.loads((root / "manifests/registry.ollama.ai/library" / repo / tag).read_text())
    for layer in m["layers"]:
        if layer["mediaType"] == "application/vnd.ollama.image.model":
            return root / "blobs" / layer["digest"].replace(":", "-")
    raise FileNotFoundError(f"{name} has no GGUF model layer")


if __name__ == "__main__":
    import collections
    import sys

    g = GGUF(ollama_model(sys.argv[1] if len(sys.argv) > 1 else "llama3.2:1b"))
    print(g.path, "gguf v", g.version)
    for k, v in g.meta.items():
        if isinstance(v, list):
            v = f"[{len(v)} items] {v[:3]}..."
        print(f"  {k}: {v}")
    print(collections.Counter(g.type_name(n) for n in g.tensors))
    for n in list(g.tensors)[:12]:
        print(f"  {n}: {g.tensors[n][0]} {g.type_name(n)}")

File: crates/tile-rt/src/gguf.rs
//! A GGUF reader: metadata, the tensor directory, and raw tensor bytes.
//!
//! Only what a Llama checkpoint needs. Values are parsed (arrays of strings
//! included, because the tokenizer lives in them and must be skipped
//! correctly), tensors are handed out as byte slices of the file — the
//! caller decides how to repack them for the kernels.

use std::collections::HashMap;
use std::path::{Path, PathBuf};

#[derive(Clone, Debug, PartialEq)]
pub enum Value {
    Int(i64),
    Float(f64),
    Bool(bool),
    Str(String),
    Array(Vec<Value>),
}

impl Value {
    pub fn as_int(&self) -> Option<i64> {
        match self {
            Value::Int(i) => Some(*i),
            _ => None,
        }
    }

    pub fn as_float(&self) -> Option<f64> {
        match self {
            Value::Float(f) => Some(*f),
            Value::Int(i) => Some(*i as f64),
            _ => None,
        }
    }
}

/// ggml tensor types this reader knows the block layout of.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum GgmlType {
    F32,
    F16,
    Q8_0,
    Other(u32),
}

impl GgmlType {
    fn from_id(id: u32) -> GgmlType {
        match id {
            0 => GgmlType::F32,
            1 => GgmlType::F16,
            8 => GgmlType::Q8_0,
            x => GgmlType::Other(x),
        }
    }

    /// (values per block, bytes per block)
    fn block(self) -> Option<(usize, usize)> {
        match self {
            GgmlType::F32 => Some((1, 4)),
            GgmlType::F16 => Some((1, 2)),
            GgmlType::Q8_0 => Some((32, 34)),
            GgmlType::Other(_) => None,
        }
    }
}

#[derive(Clone, Debug)]
pub struct TensorInfo {
    /// ggml order: `dims[0]` is the contiguous dimension.
    pub dims: Vec<usize>,
    pub ty: GgmlType,
    offset: usize,
}

impl TensorInfo {
    pub fn elems(&self) -> usize {
        self.dims.iter().product()
    }
}

pub struct Gguf {
    pub path: PathBuf,
    pub meta: HashMap<String, Value>,
    pub tensors: HashMap<String, TensorInfo>,
    bytes: Vec<u8>,
    data: usize,
}

struct Cursor<'a> {
    b: &'a [u8],
    pos: usize,
}

impl Cursor<'_> {
    fn take(&mut self, n: usize) -> Result<&[u8], String> {
        let s = self
            .b
            .get(self.pos..self.pos + n)
            .ok_or("truncated GGUF file")?;
        self.pos += n;
        Ok(s)
    }

    fn u32(&mut self) -> Result<u32, String> {
        Ok(u32::from_le_bytes(self.take(4)?.try_into().unwrap()))
    }

    fn u64(&mut self) -> Result<u64, String> {
        Ok(u64::from_le_bytes(self.take(8)?.try_into().unwrap()))
    }

    fn string(&mut self) -> Result<String, String> {
        let n = self.u64()? as usize;
        Ok(String::from_utf8_lossy(self.take(n)?).into_owned())
    }

    fn value(&mut self, ty: u32) -> Result<Value, String> {
        let v = |b: &[u8]| {
            let mut a = [0u8; 8];
            a[..b.len()].copy_from_slice(b);
            a
        };
        Ok(match ty {
            0 => Value::Int(self.take(1)?[0] as i64),
            1 => Value::Int(self.take(1)?[0] as i8 as i64),
            2 => Value::Int(u16::from_le_bytes(self.take(2)?.try_into().unwrap()) as i64),
            3 => Value::Int(i16::from_le_bytes(self.take(2)?.try_into().unwrap()) as i64),
            4 => Value::Int(self.u32()? as i64),
            5 => Value::Int(i32::from_le_bytes(self.take(4)?.try_into().unwrap()) as i64),
            6 => Value::Float(f32::from_le_bytes(self.take(4)?.try_into().unwrap()) as f64),
            7 => Value::Bool(self.take(1)?[0] != 0),
            8 => Value::Str(self.string()?),
            9 => {
                let et = self.u32()?;
                let n = self.u64()? as usize;
                let mut xs = Vec::with_capacity(n.min(1 << 20));
                for _ in 0..n {
                    xs.push(self.value(et)?);
                }
                Value::Array(xs)
            }
            10 => Value::Int(u64::from_le_bytes(v(self.take(8)?)) as i64),
            11 => Value::Int(i64::from_le_bytes(v(self.take(8)?))),
            12 => Value::Float(f64::from_le_bytes(v(self.take(8)?))),
            t => return Err(format!("unknown GGUF value type {t}")),
        })
    }
}

impl Gguf {
    pub fn open(path: &Path) -> Result<Gguf, String> {
        let bytes = std::fs::read(path).map_err(|e| format!("{}: {e}", path.display()))?;
        let mut c = Cursor { b: &bytes, pos: 0 };
        if c.take(4)? != b"GGUF" {
            return Err(format!("{} is not a GGUF file", path.display()));
        }
        let version = c.u32()?;
        if version < 2 {
            return Err(format!("GGUF version {version} is too old"));
        }
        let n_tensors = c.u64()? as usize;
        let n_kv = c.u64()? as usize;
        let mut meta = HashMap::new();
        for _ in 0..n_kv {
            let k = c.string()?;
            let t = c.u32()?;
            meta.insert(k, c.value(t)?);
        }
        let mut tensors = HashMap::new();
        for _ in 0..n_tensors {
            let name = c.string()?;
            let nd = c.u32()? as usize;
            let dims = (0..nd)
                .map(|_| c.u64().map(|d| d as usize))
                .collect::<Result<_, _>>()?;
            let ty = GgmlType::from_id(c.u32()?);
            let offset = c.u64()? as usize;
            tensors.insert(name, TensorInfo { dims, ty, offset });
        }
        let align = meta
            .get("general.alignment")
            .and_then(Value::as_int)
            .unwrap_or(32) as usize;
        let data = c.pos.next_multiple_of(align);
        Ok(Gguf {
            path: path.to_path_buf(),
            meta,
            tensors,
            bytes,
            data,
        })
    }

    pub fn int(&self, key: &str) -> Result<i64, String> {
        self.meta
            .get(key)
            .and_then(Value::as_int)
            .ok_or_else(|| format!("GGUF metadata `{key}` missing or not an integer"))
    }

    pub fn float(&self, key: &str) -> Result<f64, String> {
        self.meta
            .get(key)
            .and_then(Value::as_float)
            .ok_or_else(|| format!("GGUF metadata `{key}` missing or not a number"))
    }

    pub fn info(&self, name: &str) -> Result<&TensorInfo, String> {
        self.tensors
            .get(name)
            .ok_or_else(|| format!("tensor `{name}` not in {}", self.path.display()))
    }

    /// The tensor's bytes, exactly as stored.
    pub fn raw(&self, name: &str) -> Result<(&TensorInfo, &[u8]), String> {
        let t = self.info(name)?;
        let (per, size) =
            t.ty.block()
                .ok_or_else(|| format!("tensor `{name}` has unsupported type {:?}", t.ty))?;
        let n = t.elems() / per * size;
        let start = self.data + t.offset;
        let b = self
            .bytes
            .get(start..start + n)
            .ok_or_else(|| format!("tensor `{name}` runs past the end of the file"))?;
        Ok((t, b))
    }

    /// An F32 tensor's values.
    pub fn f32s(&self, name: &str) -> Result<Vec<f32>, String> {
        let (t, b) = self.raw(name)?;
        if t.ty != GgmlType::F32 {
            return Err(format!("tensor `{name}` is {:?}, expected F32", t.ty));
        }
        let (words, _) = b.as_chunks::<4>();
        Ok(words.iter().map(|w| f32::from_le_bytes(*w)).collect())
    }
}

/// The GGUF blob behind an Ollama model tag such as `llama3.2:1b`, found
/// through Ollama's manifest (`$OLLAMA_MODELS`, else `~/.ollama/models`).
pub fn ollama_model(tag: &str) -> Result<PathBuf, String> {
    let root = std::env::var_os("OLLAMA_MODELS")
        .map(PathBuf::from)
        .or_else(|| std::env::var_os("HOME").map(|h| PathBuf::from(h).join(".ollama/models")))
        .ok_or("cannot locate the Ollama model store")?;
    let (repo, t) = tag.split_once(':').unwrap_or((tag, "latest"));
    let manifest = root
        .join("manifests/registry.ollama.ai/library")
        .join(repo)
        .join(t);
    let text = std::fs::read_to_string(&manifest).map_err(|e| {
        format!(
            "no Ollama manifest for {tag} at {}: {e}",
            manifest.display()
        )
    })?;
    // The manifest is small JSON; find the model layer's digest without a
    // JSON dependency.
    let key = "application/vnd.ollama.image.model";
    let at = text
        .find(key)
        .ok_or_else(|| format!("{tag} has no GGUF model layer"))?;
    let rest = &text[at..];
    let d = rest.find("sha256:").ok_or("manifest layer has no digest")?;
    let digest: String = rest[d..]
        .chars()
        .take_while(|c| c.is_ascii_alphanumeric() || *c == ':')
        .collect();
    Ok(root.join("blobs").join(digest.replace(':', "-")))
}
